File: audience/src/main/java/br/com/concretesolutions/audience/core/actor/ActorRegistry.java
package br.com.concretesolutions.audience.core.actor;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;

import br.com.concretesolutions.audience.core.Director;
import br.com.concretesolutions.audience.core.exception.TragedyException;

public final class ActorRegistry {

    private static final String OLD_ACTOR_KEY =
            "br.com.concretesolutions.audience.core.actor.Audience.OLD_ACTOR_KEY";

    private final Map<String, ActorStorage> storage = new HashMap<>(100);
    private final Map<String, String> reboundPaths = new HashMap<>();

    @NonNull
    public ActorStorage storageForRef(@NonNull ActorRef ref) {
        return storageForRef(ref.ref);
    }

    @NonNull
    public ActorStorage storageForRef(@NonNull String ref) {

        if (!storage.containsKey(ref) && reboundPaths.containsKey(ref)) {
            return storage.get(reboundPaths.get(ref));
        }

        if (storage.containsKey(ref)) {
            return storage.get(ref);
        }

        throw new TragedyException("No actor registered");
    }

    public ActorRegistry enrollSingleton(@NonNull SingletonActor actor) {
        return innerEnroll(actor, actor.getClass().getCanonicalName());
    }

    public ActorRegistry enroll(@NonNull Actor actor) {
        return innerEnroll(actor, generateKey(actor));
    }

    private ActorRegistry innerEnroll(@NonNull final Actor actor, @NonNull final String key) {
        storage.put(key, new ActorStorage());
        actor.onActorRegistered(new ActorRef(key));
        return this;
    }

    public void enroll(@NonNull Actor actor, @Nullable Bundle savedInstanceState) {

        enroll(actor);

        final boolean hasOldKey = savedInstanceState != null
                && savedInstanceState.containsKey(OLD_ACTOR_KEY);

        if (hasOldKey) {
            rebindPath(savedInstanceState.getString(OLD_ACTOR_KEY), actor);
        }
    }

    public void take5(Actor actor, Bundle outState) {
        outState.putString(OLD_ACTOR_KEY, actorRefForTarget(actor).ref);
    }

    @NonNull
    public ActorRef actorRefForTarget(@NonNull Actor target) {

        final String targetKey = generateKey(target);

        if (storage.containsKey(targetKey))
            return new ActorRef(targetKey);

        throw new IllegalArgumentException("No actor registered with this key");
    }

    @NonNull
    public ActorRef actorRefForTarget(@NonNull Class<? extends Actor> target) {

        final String targetKey = target.getCanonicalName();

        if (storage.containsKey(targetKey))
            return new ActorRef(targetKey);

        throw new IllegalArgumentException("No actor registered with this key");
    }

    public RuleRegistry toRuleRegistry() {
        return Director.getRuleRegistry();
    }

    void clean(ActorRef actorRef) {
        storage.remove(actorRef.ref);
    }

    void rebindPath(String oldKey, Actor actor) {
        reboundPaths.put(oldKey, generateKey(actor));
    }

    void clear() {
        storage.clear();
        reboundPaths.clear();
    }

    private static String generateKey(Actor actor) {
        return actor.getClass().getCanonicalName()
                + System.identityHashCode(actor);
    }

}

File: audience-core/src/main/kotlin/io/github/victorolinasc/audience/core/actor/ActorRegistry.kt
package io.github.victorolinasc.audience.core.actor

import android.os.Bundle

import java.util.HashMap

import io.github.victorolinasc.audience.core.Director
import io.github.victorolinasc.audience.core.exception.TragedyException

class ActorRegistry {

    private val storage = HashMap<String, ActorStorage>(100)
    private val reboundPaths = HashMap<String, String>()

    fun storageForRef(ref: ActorRef) = storageForRef(ref.ref)

    fun storageForRef(ref: String): ActorStorage {

        if (!storage.containsKey(ref) && reboundPaths.containsKey(ref)) {
            return storage[reboundPaths[ref]]!!
        }

        if (storage.containsKey(ref)) {
            return storage[ref]!!
        }

        throw TragedyException("No actor registered")
    }

    fun enrollSingleton(actor: SingletonActor) = innerEnroll(actor, actor.javaClass.canonicalName)

    fun enroll(actor: Actor) = innerEnroll(actor, generateKey(actor))

    fun enroll(actor: Actor, savedInstanceState: Bundle?) {

        enroll(actor)

        val hasOldKey = savedInstanceState != null && savedInstanceState.containsKey(OLD_ACTOR_KEY)

        if (hasOldKey) {
            rebindPath(savedInstanceState!!.getString(OLD_ACTOR_KEY), actor)
        }
    }

    fun take5(actor: Actor, outState: Bundle) {
        outState.putString(OLD_ACTOR_KEY, actorRefForTarget(actor).ref)
    }

    fun actorRefForTarget(target: Actor): ActorRef {

        val targetKey = generateKey(target)

        if (storage.containsKey(targetKey))
            return ActorRef(targetKey)

        throw IllegalArgumentException("No actor registered with this key")
    }

    fun actorRefForTarget(target: Class<out Actor>): ActorRef {

        val targetKey = target.canonicalName

        if (storage.containsKey(targetKey))
            return ActorRef(targetKey)

        throw IllegalArgumentException("No actor registered with this key")
    }

    fun toRuleRegistry() = Director.system.ruleRegistry

    internal fun clean(actorRef: ActorRef) {
        storage.remove(actorRef.ref)
    }

    internal fun rebindPath(oldKey: String, actor: Actor) {
        reboundPaths[oldKey] = generateKey(actor)
    }

    internal fun clear() {
        storage.clear()
        reboundPaths.clear()
    }

    private fun innerEnroll(actor: Actor, key: String): ActorRegistry {
        storage[key] = ActorStorage()
        actor.onActorRegistered(ActorRef(key))
        return this
    }

    companion object {

        private const val OLD_ACTOR_KEY =
                "io.github.victorolinasc.audience.core.actor.Audience.OLD_ACTOR_KEY"

        private fun generateKey(actor: Actor) =
                actor.javaClass.canonicalName + System.identityHashCode(actor)
    }

}
